pkg/registry/registry.go

```diff
@@ -54,9 +60,124 @@ func (o RegistryClient) DownloadFileInMemory(params dfutil.HTTPRequestParams) ([
+		err = removeDirectoryContents(contextDir, o.fsys)
+		if err != nil {
+			return fmt.Errorf("failed to delete contents of the current directory; cause %w", err)
+		}
+		return util.CopyDirWithFS(starterProjectTmpDir, contextDir, o.fsys)
+	}
+
+	// Case 2: If there is no devfile, and there is no conflict between the contents of contextDir and starterproject, then copy the contents of the starterproject into contextDir.
+	// Case 3: If there is no devfile, and there is conflict between the contents of contextDir and starterproject, copy contents of starterproject into a dir named CONFLICT_STARTER_PROJECT; warn about this
+	var conflictingFiles []string
+	conflictingFiles, err = getConflictingFiles(starterProjectTmpDir, contextDir, o.fsys)
+	if err != nil {
+		return err
+	}
+
+	// Case 2
+	if len(conflictingFiles) == 0 {
+		return util.CopyDirWithFS(starterProjectTmpDir, contextDir, o.fsys)
+	}
+
+	// Case 3
+	conflictingDirPath := filepath.Join(contextDir, CONFLICT_DIR_NAME)
+	err = o.fsys.MkdirAll(conflictingDirPath, 0750)
+	if err != nil {
+		return err
+	}
+
+	err = util.CopyDirWithFS(starterProjectTmpDir, conflictingDirPath, o.fsys)
+	if err != nil {
+		return err
+	}
+	fmt.Println()
+	log.Warningf("There are conflicting files (%s) between starter project and the current directory, hence the starter project has been copied to %s", strings.Join(conflictingFiles, ", "), conflictingDirPath)
+
+	return nil
+}
+
+// removeDirectoryContents attempts to remove dir contents without deleting the directory itself, unlike os.RemoveAll() method
+func removeDirectoryContents(path string, fsys filesystem.Filesystem) error {
+	dir, err := fsys.ReadDir(path)
+	if err != nil {
+		return err
+	}
+	for _, f := range dir {
+		// a bit of cheating by using absolute file name to make sure this works with a fake filesystem, especially a memory map which is used by our unit tests
+		// memorymap's Name() method trims the full path and returns just the file name, which then becomes impossible to find by the RemoveAll method that looks for prefix
+		// See: https://github.com/redhat-developer/odo/blob/d717421494f746a5cb12da135f561d12750935f3/vendor/github.com/spf13/afero/memmap.go#L282
+		absFileName := filepath.Join(path, f.Name())
+		err = fsys.RemoveAll(absFileName)
+		if err != nil {
+			return fmt.Errorf("failed to remove %s; cause: %w", absFileName, err)
+		}
+	}
+
+	return nil
+}
+
+// getConflictingFiles fetches the contents of the two directories in question and compares them to check for conflicting files.
+// it returns a list of conflicting files (if any) along with an error (if any).
+func getConflictingFiles(spDir, contextDir string, fsys filesystem.Filesystem) (conflictingFiles []string, err error) {
+	var (
+		contextDirMap = map[string]struct{}{}
+	)
+	// walk through the contextDir, trim the file path from the file name and append it to a map
+	err = fsys.Walk(contextDir, func(path string, info fs.FileInfo, err error) error {
```

Comment:
From my understanding of the documentation of the [`WalkFunc` type](https://pkg.go.dev/path/filepath#WalkFunc), `Walk` might call this handler with a non-nil error:

> The err argument reports an error related to path, signaling that Walk will not walk into that directory. The function can decide how to handle that error; as described earlier, returning the error will cause Walk to stop walking the entire tree.

So I'd suggest returning the error, if any, earlier in this handler.